Fixed #24715 -- Renamed Oracle Sequence and Triggers when renaming table.

johnrtipton opened the pull request:
When the migration renames an Oracle table, it does not currently rename the sequences and triggers.  This renames them also if there is an auto_field.

timgraham commented:
Hi, it looks like you've sent a pull request without filing a Trac ticket. Please [file a ticket](https://code.djangoproject.com/newticket) to suggest changes and see our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) for more guidelines. Thanks!

At first glance, I think some regression tests for this will be required. Probably in `tests/schema/`

johnrtipton commented:
Thanks, I've created Ticket #24715:
https://code.djangoproject.com/ticket/24715

MarkusH commented:
Do you mind moving the inline SQL code to class-level variables as done in lines 13ff.

MarkusH commented:
In case you don't receive notifications from Trac on your ticket. Can you please add tests and mention the fix in the 1.8.1 and 1.7.8 release notes (or 1.8.2/1.7.9 in case those version are already released by then). Thank you :smile:

johnrtipton commented:
I've moved the SQL to class variables and added release notes.  I have an idea how to modify the test, but I'm not sure how to run it.  Can you tell me how I can run the django.tests.schema.test.test_db_table() test?

timgraham commented:
https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/unit-tests/ should help with running tests.

If you could rebase your branch and squash commits that would also be great.

timgraham commented on docs/releases/1.7.8.txt:
```diff
@@ -13,3 +13,6 @@ Django 1.7.8 fixes:
 * Prevented the loss of ``null``/``not null`` column properties during field
   alteration of MySQL databases (:ticket:`24595`).
+
+* Rename triggers and sequences when renaming an Oracle table using
```
suggested wording:
```
Added renaming of Oracle triggers and sequences when using ``RenameModel`` (:ticket:`24715`).
```

johnrtipton commented:
I'm working on writing the test.

I noticed that running the schema tests with and without my patch against an Oracle database results in the following.

...........FF..s.F..sFF.....EF..E............FFF...E..EE
FAILED (failures=9, errors=5, skipped=2)

I guess it's known that these tests fail against an Oracle database.

MarkusH commented:
Actually no. According to our CI all tests pass on Oracle. What Oracle version do you use? Which tests are failing specifically?

On April 30, 2015 8:54:17 PM GMT+02:00, johnrtipton <notifications@github.com> wrote:
>I'm working on writing the test.
>
>I noticed that running the schema tests with and without my patch
>against an Oracle database results in the following.
>
>...........FF..s.F..sFF.....EF..E............FFF...E..EE
>FAILED (failures=9, errors=5, skipped=2)
>
>I guess it's known that these tests fail against an Oracle database.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/django/django/pull/4567#issuecomment-97925536

johnrtipton commented:
Oracle 12c:

$ PYTHONPATH=..:$PYTHONPATH ./runtests.py --settings=settings schema
Testing against Django installed in '/Users/tip/Dropbox/github/django-oracle-fix/django'
Creating test database for alias 'default'...
Creating test user...
...........FF..s.F..sFF.....EF..E............FFF...E..EE
======================================================================
ERROR: test_fk (schema.tests.SchemaTests)
Tests that creating tables out of FK order, then repointing, works
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 162, in test_fk
    editor.alter_field(Book, old_field, new_field, strict=True)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/oracle/schema.py", line 78, in alter_field
    super(DatabaseSchemaEditor, self).alter_field(model, old_field, new_field, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 487, in alter_field
    old_db_params, new_db_params, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 501, in _alter_field
    old_field.column,
ValueError: Found wrong number (0) of foreign key constraints for schema_book.author_id

======================================================================
ERROR: test_indexes (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 1217, in test_indexes
    editor.alter_field(BookWithSlug, new_field3, new_field4, strict=True)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/oracle/schema.py", line 78, in alter_field
    super(DatabaseSchemaEditor, self).alter_field(model, old_field, new_field, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 487, in alter_field
    old_db_params, new_db_params, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 514, in _alter_field
    old_field.column,
ValueError: Found wrong number (0) of unique constraints for schema_book.slug

======================================================================
ERROR: test_primary_key (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 1242, in test_primary_key
    editor.alter_field(Tag, old_field, new_field)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/oracle/schema.py", line 78, in alter_field
    super(DatabaseSchemaEditor, self).alter_field(model, old_field, new_field, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 487, in alter_field
    old_db_params, new_db_params, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 700, in _alter_field
    "columns": self.quote_name(new_field.column),
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 110, in execute
    cursor.execute(sql, params)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/utils.py", line 95, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/oracle/base.py", line 478, in execute
    return self.cursor.execute(query, self._param_generator(params))
DatabaseError: ORA-02261: such unique or primary key already exists in the table


======================================================================
ERROR: test_unique (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 1033, in test_unique
    editor.alter_field(Tag, old_field, new_field, strict=True)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/oracle/schema.py", line 78, in alter_field
    super(DatabaseSchemaEditor, self).alter_field(model, old_field, new_field, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 487, in alter_field
    old_db_params, new_db_params, strict)
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 514, in _alter_field
    old_field.column,
ValueError: Found wrong number (0) of unique constraints for schema_tag.slug

======================================================================
ERROR: test_unique_together (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 1072, in test_unique_together
    editor.alter_unique_together(UniqueTest, UniqueTest._meta.unique_together, [])
  File "/Users/tip/Dropbox/github/django-oracle-fix/django/db/backends/base/schema.py", line 324, in alter_unique_together
    ", ".join(columns),
ValueError: Found wrong number (0) of constraints for schema_uniquetest(year, slug)

======================================================================
FAIL: test_alter_fk (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 562, in test_alter_fk
    self.fail("No FK constraint for author_id found")
AssertionError: No FK constraint for author_id found

======================================================================
FAIL: test_alter_fk_to_o2o (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 691, in test_alter_fk_to_o2o
    self.assertTrue(author_is_fk, "No FK constraint for author_id found")
AssertionError: No FK constraint for author_id found

======================================================================
FAIL: test_alter_o2o_to_fk (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 645, in test_alter_o2o_to_fk
    self.assertTrue(author_is_fk, "No FK constraint for author_id found")
AssertionError: No FK constraint for author_id found

======================================================================
FAIL: test_alter_to_fk (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 618, in test_alter_to_fk
    self.fail("No FK constraint for author_id found")
AssertionError: No FK constraint for author_id found

======================================================================
FAIL: test_check_constraints (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 995, in test_check_constraints
    self.fail("No check constraint for height found")
AssertionError: No check constraint for height found

======================================================================
FAIL: test_fk_db_constraint (schema.tests.SchemaTests)
Tests that the db_constraint parameter is respected
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 211, in test_fk_db_constraint
    self.fail("No FK constraint for tag_id found")
AssertionError: No FK constraint for tag_id found

======================================================================
FAIL: test_m2m_repoint (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 973, in test_m2m_repoint
    self._test_m2m_repoint(ManyToManyField)
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 953, in _test_m2m_repoint
    self.fail("No FK constraint for tagm2mtest_id found")
AssertionError: No FK constraint for tagm2mtest_id found

======================================================================
FAIL: test_m2m_repoint_custom (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 976, in test_m2m_repoint_custom
    self._test_m2m_repoint(CustomManyToManyField)
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 953, in _test_m2m_repoint
    self.fail("No FK constraint for tagm2mtest_id found")
AssertionError: No FK constraint for tagm2mtest_id found

======================================================================
FAIL: test_m2m_repoint_inherited (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 979, in test_m2m_repoint_inherited
    self._test_m2m_repoint(InheritedManyToManyField)
  File "/Users/tip/Dropbox/github/django-oracle-fix/tests/schema/tests.py", line 953, in _test_m2m_repoint
    self.fail("No FK constraint for tagm2mtest_id found")
AssertionError: No FK constraint for tagm2mtest_id found

----------------------------------------------------------------------
Ran 56 tests in 186.884s

FAILED (failures=9, errors=5, skipped=2)
Destroying test database for alias 'default'...
Destroying test user...
Destroying test database tables...

johnrtipton commented:
I'm having trouble creating a record in the test, it never sends the insert to the database.

I'm modifying the test_db_table and adding a objects.create statement just after a print "HELLO" statement.  I'm printing the SQL to the screen and I see this:

CREATE TABLE "schema_author" ("id" integer NOT NULL PRIMARY KEY AUTOINCREMENT, "name" varchar(255) NOT NULL, "height" integer unsigned NULL)
ALTER TABLE "schema_author" RENAME TO "schema_otherauthor"
HELLO
ALTER TABLE "schema_otherauthor" RENAME TO "schema_author"

Is there something special I need to do to commit this new record?

shaib commented:
You shouldn't need to. Your runs look funny though -- the Django test suite requires two databases to be defined ("default" and "other"), and you seem to have only one. Perhaps there is some issue with routers which breaks your tests and the saving of your records.

johnrtipton commented:
Thanks shaib.  I've not got other defined, which is probably my problem.

MarkusH commented:
Test suite passes here on Oracle 12c ("Oracle DB Developer VM) too.

johnrtipton commented:
I'm making changes to the test_db_alter test to try to reset the sequence after we rename a table.  This is how I originally ran into the problem.  But when I add the following line to the test, it seems to have no effect.  I'm printing the SQL and it never tries to reset the sequence.  Do I not have access to this in the tests, or am I doing something wrong?

`connection.ops.sequence_reset_sql(no_style(), [Author, ])`

Here is the full method:

    def test_db_table(self):
        """
        Tests renaming of the table
        """
        # Create the table
        with connection.schema_editor() as editor:
            editor.create_model(Author)
        # Ensure the table is there to begin with
        columns = self.column_classes(Author)
        self.assertEqual(columns['name'][0], "CharField")
        # Alter the table
        with connection.schema_editor() as editor:
            editor.alter_db_table(Author, "schema_author", "schema_otherauthor")
        # Ensure the table is there afterwards
        Author._meta.db_table = "schema_otherauthor"
        columns = self.column_classes(Author)
        self.assertEqual(columns['name'][0], "CharField")

        # Reset sequence
        connection.ops.sequence_reset_sql(no_style(), [Author, ])

        # Alter the table again
        with connection.schema_editor() as editor:
            editor.alter_db_table(Author, "schema_otherauthor", "schema_author")
        # Ensure the table is still there
        Author._meta.db_table = "schema_author"
        columns = self.column_classes(Author)
        self.assertEqual(columns['name'][0], "CharField")

timgraham commented:
`sequence_reset_sql()` returns a SQL statement -- you still need to execute it yourself.

shaib commented:
On Friday 01 May 2015 17:35:20 johnrtipton wrote:
> I'm making changes to the test_db_alter test to try to reset the sequence
> after we rename a table.  This is how I originally ran into the problem. 

I had been thinking that this bug is a little suspicious. The fact that the 
trigger and sequence name shouldn't really matter, as long as they are still 
connected to the right table. If the names are computed from the model (and 
not kept in memory) then a future migration will fail. But resetting the 
sequence is not an operation that should even happen in production. If the 
only failure involves an operation for testing, I am no longer sure that the 
fix should be backported.

To prove that backporting is needed, I think we need a test that fails on one 
of these scenarios:

1) Rename table, then save an object.

2) Rename table, then rename table again (to a name different from the 
orignal), then save an object.

johnrtipton commented:
In Django 1.7.7 if you try to do a manage flush after renaming a table it will fail.  This is why I filled the bug report and patch.

johnrtipton commented:
Sorry for all the confusion, this is my first PR, so I'm new at this.  I've added the test, and I noticed that supports_sequence_reset is set to False for Oracle.  But it seems that manage flush still tries to reset the sequence.  Is this a legacy setting that should be changed, or should manage flush check for this?

Thanks for all the guidance you all have provided.

timgraham commented:
`supports_sequence_reset` controls sequence reset behavior during testing - not during flushing in general, see `django/test/testcases.py: if conn.features.supports_sequence_reset:`

johnrtipton commented:
Ok, it looks like sequence_reset_sql is on db.backends.base.operations and I don't have to check to see if the database supports sequence resets.

I think the patch is finished, let me know if you see further improvements that should be made.

MarkusH commented:
Can you rebase your test onto current master and move the release notes to 1.7.9 and 1.8.2 repesctively, please. We released 1.7.8 and 1.8.1 a couple of hours ago. Thanks :smile:

johnrtipton commented:
Rebased onto current master and changed release notes.

timgraham commented:
buildbot, test on oracle.

timgraham commented on docs/releases/1.7.9.txt:
```diff
@@ -0,0 +1,10 @@
+Bugfixes
+========
+
+* Added renaming of Oracle triggers and sequences when using ``RenameModel`` (:ticket:`24715`).
```
please wrap docs at 79 characters and also add this document to docs/releases/index.txt to fix the build failure.

timgraham commented:
Please see the "oracle" build for some failures.

johnrtipton commented:
I've fixed the migrations.test_operations.OperationTests.test_alter_model_table_noop test.

The following two tests fail before my patch is applied:
schema.tests.SchemaTests.test_m2m_repoint_custom
schema.tests.SchemaTests.test_m2m_repoint_inherited

timgraham commented:
Those tests don't fail for me on master (nor on the CI server). Could you double check?

johnrtipton commented:
Here is my session, pulling from the Django repository and running the test.  Unless there is something wrong with my environment, it doesn't pass against the official master.  I believe this is the same place the build was failing in Jenkins.

```
Johns-iMac:django-patch tip$ git clone https://github.com/django/django.git
Cloning into 'django'...
remote: Counting objects: 317458, done.
remote: Compressing objects: 100% (55/55), done.
remote: Total 317458 (delta 23), reused 0 (delta 0), pack-reused 317403
Receiving objects: 100% (317458/317458), 128.85 MiB | 1.36 MiB/s, done.
Resolving deltas: 100% (227138/227138), done.
Checking connectivity... done.
Johns-iMac:django-patch tip$ cd django/
Johns-iMac:django tip$ cp ../../settings.py ./
Johns-iMac:django tip$ cd tests
Johns-iMac:tests tip$ PYTHONPATH=..:$PYTHONPATH ./runtests.py --settings=settings schema.tests.SchemaTests.test_m2m_repoint_inherited
Testing against Django installed in '/Users/tip/Dropbox/github/django-patch/django/django'
Creating test database for alias 'default'...
Creating test user...
Creating test database for alias 'other'...
Creating test user...
F
======================================================================
FAIL: test_m2m_repoint_inherited (schema.tests.SchemaTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/tip/Dropbox/github/django-patch/django/tests/schema/tests.py", line 979, in test_m2m_repoint_inherited
    self._test_m2m_repoint(InheritedManyToManyField)
  File "/Users/tip/Dropbox/github/django-patch/django/tests/schema/tests.py", line 953, in _test_m2m_repoint
    self.fail("No FK constraint for tagm2mtest_id found")
AssertionError: No FK constraint for tagm2mtest_id found

----------------------------------------------------------------------
Ran 1 test in 13.402s

FAILED (failures=1)
Destroying test database for alias 'default'...
Destroying test user...
Destroying test database tables...
Destroying test database for alias 'other'...
Destroying test user...
Destroying test database tables...
Johns-iMac:tests tip$ cat ../settings.py
# Django settings for docman project.
import os

DEBUG = True
TEMPLATE_DEBUG = DEBUG

PROJECT_PATH = os.path.realpath(os.path.dirname(__file__))

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': '10.*.*.*/*******',
        'USER': '******',
        'PASSWORD': '********',
   },
    'other': {
        'ENGINE': 'django.db.backends.oracle',
        'NAME': '10.*.*.*/*******',
        'USER': '*******',
        'PASSWORD': '********',
    }
}

SECRET_KEY = 'SUPERSECRET'
```

shaib commented:
Hi,

I, like Tim and Jenkins, don't see failures on master; also, for the traceback you have, the fact that `test_m2m_repoint_inherited` fails while `test_m2m_repoint` succeeds is quite surprising.

My settings differ from yours -- I have 
```python
        'OPTIONS': {
            'threaded': True,
            },
```
in my database settings, but commenting it out changed nothing.

Do you have the latest cx_oracle (5.1.3)?

timgraham commented:
Hi John, are you still stuck on this?

johnrtipton commented:
Tim, sorry I've been really busy on a big project.  I've setup environments on different machines, but these tests do not seem to pass even without my modifications.  I'm running cx_oracle 5.1.2 in my environment so that may be the cause of some of the failures.  When I get some free time I'll setup another environment with the latest cx_oracle and try again.

timgraham commented:
Please send a new PR when you have a chance, thanks!

crazyh commented:
Looks like the same issue exist for PostgreSQL.

I have created model, added RenameModel migration and see that the table (with new name) still uses old sequence.

timgraham commented:
That might be the same issue as https://code.djangoproject.com/ticket/24533, feel free to add a comment there or create a new ticket if you believe it's a different issue.

crazyh commented:
Yep, it looks that DJango migration backends just do not support sequence modification properly. Thx for link.
